### PR TITLE
 Add support for Bugzilla attachments

### PR DESCRIPTION
**Pull request by jamacku**
This PR is attempting to add support for:
- [x] Getting Attachments
- [x] Creating Attachments
- [x] Updating Attachments

Status:
- [x] Tested
- [x] Documented

---
PR also includes small bugfixes, documentation updates and refactoring.
Related to #40

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -146,3 +146,49 @@ let response = await api.updateBug(123456, {
...
+  content_type: "text/plain";
+  is_private?: false;
+});
+```
+
+Returned value is an array of IDs of the newly-created attachments.
+
+## Updating attachments
+
+Example of changing content type of attachment #123456:
```
Comment (Mossop) [inline README.md:189]: Is it possible to change anything other than content type with this API?

File: README.md [hunk 2]
```diff
@@ -146,3 +146,49 @@ let response = await api.updateBug(123456, {
...
+```
+
+Return value is array of Attachment objects.
+
+## Creating attachments
+
+```javascript
+let attachment = await api.createAttachment(123456, {
+  ids: [123456, 123457];
+  data: "base64-content";
```
Comment (Mossop) [inline README.md]: Do you think it would be more useful to callers to allow passing a Buffer or an ArrayBuffer and then convert to base-64 internally?

**Review by Mossop — CHANGES_REQUESTED** [review]
Couple of mainly questions here otherwise I think this looks pretty good.

File: README.md (hunk 1 above)
Comment (jamacku) [reply README.md:189]: In theory it should be possible to change more than only content type - [docs](https://bugzilla.readthedocs.io/en/latest/api/core/v1/attachment.html#update-attachment). 

It is just an example. Should I adjust the wording of README to include more options?

File: README.md (hunk 2 above)
Comment (jamacku) [reply README.md]: It sounds like a reasonable thing to do, I'll update PR. Thanks for the suggestion.

**Comment by Mossop** [comment]
You'll probably need to rebase this now that #52 is fixed.

File: README.md (hunk 2 above)
Comment (jamacku) [reply README.md]: @Mossop I updated PR, please have a look - https://github.com/Mossop/bugzilla-ts/pull/81/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R284-R286

File: src/types.ts [hunk 3]
```diff
@@ -399,3 +396,112 @@ export interface UpdatedBugTemplate {
 export const UpdatedBugTemplateSpec: ObjectSpec<UpdatedBugTemplate> = {
   bugs: array(object(UpdatedBugSpec)),
 };
+
+export interface Attachment {
+  data: base64;
```
Comment (Mossop) [inline src/types.ts]: I think by the same token we should decode the base64 and convert it into a Buffer.

File: src/validators.ts [hunk 4]
```diff
@@ -85,6 +85,7 @@ export const boolean = typedValidator<boolean>("boolean");
 export const int = typedValidator<number>("number");
 export const double = typedValidator<number>("number");
 export const string = typedValidator<string>("string");
+export const base64 = typedValidator<string>("string");
```
Comment (Mossop) [inline src/validators.ts]: Instead of this the validator should decode the data and return a buffer here.

**Review by Mossop — CHANGES_REQUESTED** [review]
Can you try adding some tests using the new bits I added in #91? You should be able to use the example from `bugs.test.ts` to create a bug then create, update and list its attachments. If you run into problems with how the tests are running don't worry too much about it just let me know where you got stuck.

**Comment by jamacku** [comment]
> Can you try adding some tests using the new bits I added in #91? You should be able to use the example from `bugs.test.ts` to create a bug then create, update and list its attachments. If you run into problems with how the tests are running don't worry too much about it just let me know where you got stuck.

Thank you, I plan to add some tests via #92. I'll add relevant tests to this PR as well.

File: itest/test/attachments.test.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+      // flags: [{ name: "new-flag", status: "?", new: true }],
+      // is_markdown: false,
+      is_obsolete: true,
+      is_patch: true,
+      // is_private: true,
+      summary: "new summary",
+    }),
+  ).resolves.toEqual([
+    {
+      changes: new Map([
```
Comment (jamacku) [inline itest/test/attachments.test.ts]: For some reason the returned `Map` is changing the order of returned values, therefore the test is failing.

Noted, that `Map` in test above is passing just fine. 

Any idea how to deal with it? @Mossop Thank you

File: src/validators.ts [hunk 6]
```diff
@@ -190,3 +191,15 @@ export function either<F, S>(
     return result;
   };
 }
+
+export function decodeBase64<T>(validator: Validator<T>): Buffer {
```
Comment (jamacku) [inline src/validators.ts]: I'm a bit lost on this validator, @Mossop do you have any suggestions? Thank you.

Comment (Mossop) [reply src/validators.ts]: You probably want something like:
```typescript
export function base64(val: any): Buffer {
  if (typeof val != "string") {
    throw new Error(`Expected a base64 encoded string but received ${repr(val)}`);
  }

  try {
    return Buffer.from(val, "base64");
  } catch (e) {
    throw new Error(`Expected a base64 encoded string but received ${repr(val)}`);
  }
}
```

File: src/types.ts (hunk 3 above)
Comment (jamacku) [reply src/types.ts]: Done

File: src/validators.ts (hunk 4 above)
Comment (jamacku) [reply src/validators.ts]: Done

File: src/validators.ts (hunk 6 above)
Comment (jamacku) [reply src/validators.ts]: Thank you, works perfectly! :+1:

**Review by Mossop — APPROVED** [review]

**Pull request merged**